### ISSUE TITLE
How to remap keys in another mode's keymap

### ISSUE THREAD

**Issue by raxod502** [jwiegley/use-package#413]
Currently I have a `use-package` form that looks like this:

```
(use-package counsel-projectile
  :bind (:map projectile-mode-map
         ("SPC" . counsel-projectile)
         ([remap projectile-find-file]
          . counsel-projectile-find-file)))
```

The problem is, the second keymapping doesn't work properly. In particular, `C-c p f`, the normal binding for `projectile-find-file`, remains bound to `projectile-find-file` instead of being bound to `counsel-projectile-find-file`. But if I reload my init-file, then it works. That is, `C-c p f` is bound to `counsel-projectile-find-file`.

I figured that this might be because `projectile` was loaded after `counsel-projectile`, and therefore my remapping was being overwritten. So I added `:after projectile` to the `counsel-projectile` `use-package` form. That fixed the problem, but now `counsel-projectile` is loaded eagerly as soon as `projectile` is loaded.

Any suggestions for remapping a key in a keymap defined by another package without eagerly loading the package whose command I'm remapping the key to?

(I can provide an `emacs -Q` compatible example, if desired.)

**Comment by jwiegley** [comment]
I would recommend expanding the macro invocation here, finding out exactly what `use-package` is doing behind the scenes.  Then, figure out what the correct sequence of commands should be, and then come back and show me the difference.  Then maybe we can think of a way to automate it correctly.  The goal for `use-package` is to always "do what makes the most sense", with an eye to preserving as much performance as possible, without incurring unexpected behavior.

**Comment by raxod502** [comment]
Before I do that, I need to ask you: what are the semantics of `:after` intended to be? That keyword is not documented, so I'm not sure if the current behavior is desired or not.

Based on the macroexpansion of the above `use-package` declaration, it looks like `:after` eagerly loads the package once the specified feature is loaded. Whereas I would have expected it to simply inhibit the loading of the package until the specified feature is loaded, but to even then wait until an autoloaded function is called.

What's your opinion?

**Comment by spatz** [comment]
I have the same problem with e.g. `python-mode-map`.
It seems to me that use-package doesn't autoload the keymap so that your binding is overridden when it is defined.

**Comment by jwiegley** [comment]
@raxod502 All that `:after foo` does is to insert an `eval-after-load 'foo` form around whatever else `use-package` would have done, so that we try to continue configuration only after `foo` has been loaded.

**Comment by raxod502** [comment]
@jwiegley Actually, I don't think that's the case. (It's what I would expect to happen, but not what actually happens.)

The following declaration:

```
(use-package foo
  :defer t
  :after bar
  :config (baz))
```

expands to this:

```
(progn
  (progn
    (eval-after-load 'bar
      '(require 'foo nil t)))
  (eval-after-load 'foo
    '(progn
       (condition-case-unless-debug err
           (baz)
         (error
          (ignore
           (display-warning 'use-package
                            (format "%s %s: %s" "foo" ":config"
                                    (error-message-string err))
                            :error))))
       t)))
```

That is, `:after bar` causes `foo` to be eagerly loaded as soon as `bar` is loaded. I think the following expansion would make more sense:

```
(progn
  (eval-after-load 'bar
    (eval-after-load 'foo
      '(progn
         (condition-case-unless-debug err
             (baz)
           (error
            (ignore
             (display-warning 'use-package
                              (format "%s %s: %s" "foo" ":config"
                                      (error-message-string err))
                              :error))))
         t))))
```

What do you think?

**Comment by kaushalmodi** [comment]
@raxod502

If the helps, I do this to for binding overrides to happen all over the place after loading counsel:

```el
(use-package counsel
  :defer t
  :init
  (progn
    ;; Do not bind the below keys to counsel commands if the user has decided
    ;; to use ido instead of ivy.
    (when (not (bound-and-true-p disable-pkg-ivy))
      (bind-keys
       :map modi-mode-map
        ("M-x"     . counsel-M-x)
        ("C-M-y"   . counsel-yank-pop)
        ("C-x C-f" . counsel-find-file)
        ("C-h v"   . counsel-describe-variable)
        ("C-h f"   . counsel-describe-function)
        ("C-h S"   . counsel-info-lookup-symbol)
        ("C-c u"   . counsel-unicode-char))
      (bind-keys
       ("M-o" . counsel-recentf))
      (bind-to-modi-map "v" #'counsel-set-variable)
      (bind-keys :map read-expression-map
        ("C-r" . counsel-expression-history)) ; useful in `eval-expression' (`M-:')
      (bind-chords
       ("JJ" . counsel-imenu)
       (";'" . counsel-M-x))
      (with-eval-after-load 'org
        (bind-key "C-c C-q" #'counsel-org-tag org-mode-map))
      (with-eval-after-load 'org-agenda
        (bind-key "C-c C-q" #'counsel-org-tag-agenda org-agenda-mode-map))))
  :config
 ;; <snip>
 ;; ..
```

I use the plain old `:init` over here instead of `:bind` to get full control on if I want the bindings to happen and where.

**Comment by raxod502** [comment]
Oh, duh. It always felt clumsy to do my bindings in `:init` or `:config`, but of course you can use `bind-key` directly to get syntax that's just as nice.

**Comment by raxod502** [comment]
Closing because the comment by @kaushalmodi solves my original problem, and the behavior/lack of documentation for `:after` is a separate issue.
